src/identity/manager.rs

```diff
@@ -44,18 +52,306 @@ impl fmt::Display for Identity {
+    async fn fetch_certificate(&mut self, id: &Identity) -> Result<tls::Certs, Error> {
+        loop {
+            let mut tx_option: Option<watch::Sender<Option<tls::Certs>>> = None;
+            let mut cache_rx = None;
+            {
+                let read_locked_cache = self.cache.read().unwrap();
+                match read_locked_cache.get(id) {
+                    None => {
+                        drop(read_locked_cache);
+                        let mut write_locked_cache = self.cache.write().unwrap();
+                        match write_locked_cache.get(id) {
+                            Some(cert_rx) => {
+                                // A different thread got here before us and is handling the fetch.
+                                // Take a copy of the receiver.
+                                cache_rx = Some(cert_rx.clone());
+                            }
+                            None => {
+                                let (tx, rx) = watch::channel(None);
+                                write_locked_cache.insert(id.clone(), rx);
+                                tx_option = Some(tx);
+                            }
+                        };
+                    }
+                    Some(cert_rcvr) => {
+                        cache_rx = Some(cert_rcvr.clone());
+                    }
+                }
+            }
+
+            // We made a transmitter, so fetch the cert and send the result.
+            if let Some(tx) = tx_option {
+                info!("Fetching cert.");
+                let certs = self.client.fetch_certificate(id).await;
+                match certs {
+                    Ok(c) => {
+                        match tx.send(Some(c.clone())) {
+                            Ok(_) => {
+                                tokio::spawn(SecretManager::refresh_handler(
+                                    id.clone(),
+                                    self.cache.clone(),
+                                    self.client.clone(),
+                                    c.get_duration_until_refresh(),
+                                    tx,
+                                ));
+                            }
+                            Err(e) => {
+                                warn!("Failed to send fetched certs on channel: {:?}", e);
+                                let mut locked_cache = self.cache.write().unwrap();
+                                locked_cache.remove(id);
+                            }
+                        }
+
+                        return Ok(c);
+                    }
+                    Err(e) => {
+                        let mut locked_cache = self.cache.write().unwrap();
+                        locked_cache.remove(id);
+                        return Err(e);
+                    }
+                }
+            }
+
+            // We found a channel receiver in the cache, so use it to get the cert.
+            loop {
```

Comment:
Why must the cache receiver send `Option`? Seems like we could just block here until the receiver gets something instead of looping. I don't see any `.send(None)`.

Reply:
L142:
`let (tx, rx) = watch::channel(None);`

That is sending None.  Without an Option we can't initialize the receiver until we have the cert.